packages/firestore/test/integration/remote/stream.test.ts

```diff
@@ -0,0 +1,231 @@
+import { EmptyCredentialsProvider } from '../../../src/api/credentials';
+import { SnapshotVersion } from '../../../src/core/snapshot_version';
+import { MutationResult } from '../../../src/model/mutation';
+import { PlatformSupport } from '../../../src/platform/platform';
+import { Datastore } from '../../../src/remote/datastore';
+import {
+  PersistentListenStream,
+  PersistentWriteStream,
+  WatchStreamListener,
+  WriteStreamListener
+} from '../../../src/remote/persistent_stream';
+import {
+  DocumentWatchChange,
+  ExistenceFilterChange,
+  WatchTargetChange
+} from '../../../src/remote/watch_change';
+import { AsyncQueue } from '../../../src/util/async_queue';
+import { Deferred } from '../../../src/util/promise';
+import { asyncIt } from '../../util/helpers';
+import { getDefaultDatabaseInfo } from '../util/helpers';
+import FirestoreError = firestore.FirestoreError;
+
+type StreamStatusCallback =
+  | 'onHandshakeComplete'
+  | 'onMutationResult'
+  | 'onWatchChange'
+  | 'onOpen'
+  | 'onClose';
+
+class StreamStatusListener implements WatchStreamListener, WriteStreamListener {
+  private pendingStates: StreamStatusCallback[] = [];
+  private pendingPromises: Deferred<StreamStatusCallback>[] = [];
+
+  private resolvePending(actualCallback: StreamStatusCallback) {
+    let pendingPromise = this.pendingPromises.shift();
+    if (pendingPromise) {
+      pendingPromise.resolve(actualCallback);
+    } else {
+      this.pendingStates.push(actualCallback);
+    }
+    return Promise.resolve();
+  }
+
+  onHandshakeComplete(): Promise<void> {
+    return this.resolvePending('onHandshakeComplete');
+  }
+
+  onMutationResult(
+    commitVersion: SnapshotVersion,
+    results: MutationResult[]
+  ): Promise<void> {
+    return this.resolvePending('onMutationResult');
+  }
+
+  onWatchChange(
+    watchChange:
+      | DocumentWatchChange
+      | WatchTargetChange
+      | ExistenceFilterChange,
+    snapshot: SnapshotVersion
+  ): Promise<void> {
+    return this.resolvePending('onWatchChange');
+  }
+
+  onOpen(): Promise<void> {
+    return this.resolvePending('onOpen');
+  }
+
+  onClose(err?: FirestoreError): Promise<void> {
+    return this.resolvePending('onClose');
+  }
+
+  /**
+   * Returns a Promise that resolves when 'expectedCallback' fires.
+   * Resolves the returned Promise immediately if there is already an
+   * unprocessed callback.
+   * Fails the test if the expected callback does not match the name of the
+   * actual callback function.
+   */
+  awaitCallback(expectedCallback: StreamStatusCallback): Promise<void> {
+    if (this.pendingStates.length > 0) {
+      expect(this.pendingStates.shift()).to.equal(expectedCallback);
+      return Promise.resolve();
+    } else {
+      const deferred = new Deferred<StreamStatusCallback>();
+      this.pendingPromises.push(deferred);
+      return deferred.promise.then(actualCallback => {
+        expect(actualCallback).to.equal(expectedCallback);
+      });
+    }
```

Comment:
This code is a bit wonky in that on failure it might throw an exception or might return a failed promise, depending on if we have pendingStates (and that the expectedCallback expectation is repeated).  Perhaps do something like:

```
let promise: Promise<StreamStatusCallback>;
if (this.pendingStates.length > 0) {
  promise = Promise.resolve(this.pendingStates.shift());
} else {
  const deferred = new Deferred<StreamStatusCallback>();
  this.pendingPromises.push(deferred);
  promise = deferred.promise;
}
return promise.then(actualCallback => {
  expect(actualCallback).to.equal(expectedCallback);
});
```

Reply:
I changed the code to always return a Promise, and to reject it if the callback type doesn't match.